usr/lib/soft_stdll/soft_specific.c

```diff
@@ -3051,3 +3075,540 @@ CK_RV token_specific_aes_cmac(STDLL_TokData_t *tokdata, CK_BYTE *message,
+    if (rc != CKR_OK || field_len != data_len) {
+        TRACE_DEVEL("ber_decode_OCTET_STRING failed\n");
+        rc = CKR_ATTRIBUTE_VALUE_INVALID;
+        goto out;
+    }
+
+    /* Check for public key without format byte */
+    privlen = (EC_GROUP_order_bits(EC_KEY_get0_group(ec_key)) + 7) / 8;
+    form  = ecpoint[0] & ~0x01;
+    if (ecpoint_len <= 2 * privlen &&
+        form != POINT_CONVERSION_COMPRESSED &&
+        form != POINT_CONVERSION_UNCOMPRESSED &&
+        form != POINT_CONVERSION_HYBRID) {
+        temp = malloc(1 + 2 * privlen);
+        if (temp == NULL) {
+            rc = CKR_HOST_MEMORY;
+            goto out;
+        }
+
+        padlen = 2 * privlen - ecpoint_len;
+        temp[0] = POINT_CONVERSION_UNCOMPRESSED;
+        memset(temp + 1, 0, padlen);
+        memcpy(temp + 1 + padlen, ecpoint, ecpoint_len);
+
+        ecpoint = temp;
+        ecpoint_len = 1 + 2 * privlen;
+    }
+
+    if (!EC_KEY_oct2key(ec_key, ecpoint, ecpoint_len, NULL)) {
+        TRACE_ERROR("EC_KEY_oct2key failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+out:
+    if (temp != NULL)
+        free(temp);
+
+    return rc;
+}
+
+static CK_RV fill_ec_key_from_privkey(EC_KEY *ec_key, const CK_BYTE *data,
+                                      CK_ULONG data_len)
+{
+    EC_POINT *point = NULL;
+    CK_RV rc = CKR_OK;
+
+    if (!EC_KEY_oct2priv(ec_key, data, data_len)) {
+        TRACE_ERROR("EC_KEY_oct2priv failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    point = EC_POINT_new(EC_KEY_get0_group(ec_key));
+    if (point == NULL) {
+        TRACE_ERROR("EC_POINT_new failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (!EC_POINT_mul(EC_KEY_get0_group(ec_key), point,
+                      EC_KEY_get0_private_key(ec_key), NULL, NULL, NULL)) {
+        TRACE_ERROR("EC_POINT_mul failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (!EC_KEY_set_public_key(ec_key, point)) {
+        TRACE_ERROR("EC_KEY_set_public_key failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+out:
+    if (point != NULL)
+        EC_POINT_free(point);
+
+    return rc;
+}
+
+static CK_RV make_ec_key_from_template(TEMPLATE *template, EC_KEY **key)
+{
+    CK_ATTRIBUTE *attr = NULL;
+    CK_OBJECT_CLASS keyclass;
+    EC_KEY *ec_key = NULL;
+    CK_RV rc;
+
+    rc = template_attribute_find(template, CKA_CLASS, &attr);
+    if (rc == FALSE) {
+        TRACE_ERROR("Could not find CKA_CLASS in the template\n");
+        rc = CKR_TEMPLATE_INCOMPLETE;
+        goto out;
+    }
+
+    keyclass = *(CK_OBJECT_CLASS *) attr->pValue;
+
+    if (!template_attribute_find(template, CKA_ECDSA_PARAMS, &attr)) {
+        TRACE_ERROR("Could not find CKA_ECDSA_PARAMS in the template\n");
+        rc = CKR_TEMPLATE_INCOMPLETE;
+        goto out;
+    }
+
+    rc = make_ec_key_from_params(attr->pValue, attr->ulValueLen, &ec_key);
+    if (rc != CKR_OK)
```

Comment:
Why no TRACE_ERROR here? Did the function already write a related trace entry?

Reply:
Yes, make_ec_key_from_params already traces all error situations.